fileserver/cook/file_server.py

```diff
@@ -0,0 +1,164 @@
+
+
+class FileServerApplication(gunicorn.app.base.BaseApplication):
+
+    def __init__(self, cook_workdir, options=None):
+        self.options = options or {}
+        self.application = app
+        global sandbox_directory
+        sandbox_directory = cook_workdir
+        super(FileServerApplication, self).__init__()
+
+    def load_config(self):
+        for key, value in self.options.items():
+            if key in self.cfg.settings and value is not None:
+                self.cfg.set(key.lower(), value)
+
+    def load(self):
+        return self.application
+
+
+def path_is_valid(path):
+    if not os.path.exists(path):
+        return False
+    normalized_path = os.path.normpath(path)
+    return normalized_path.startswith(sandbox_directory)
+
+
+@app.route('/files/download')
+@app.route('/files/download.json')
+def download():
+    path = request.args.get('path')
+    if path is None:
+        return "Expecting 'path=value' in query.\n", 400
+    if not path_is_valid(path):
+        return "", 404
+    if os.path.isdir(path):
+        return "Cannot download a directory.\n", 400
+    return send_file(path, as_attachment=True)
+
+
+@app.route('/files/read')
+@app.route('/files/read.json')
+def read():
+    path = request.args.get('path')
+    offset_param = request.args.get('offset', -1)
+    length_param = request.args.get('length', -1)
+    if path is None:
+        return "Expecting 'path=value' in query.\n", 400
+    try:
+        offset = int(offset_param)
+    except ValueError as _:
+        return f"Failed to parse offset: Failed to convert '{offset_param}' to number.\n", 400
+    if offset < -1:
+        return f"Negative offset provided: {offset_param}.\n", 400
+    try:
+        length = int(length_param)
+    except ValueError as _:
+        return f"Failed to parse length: Failed to convert '{length_param}' to number.\n", 400
+    if length < -1:
+        return f"Negative length provided: {length_param}.\n", 400
+    if not path_is_valid(path):
+        return "", 404
+    if os.path.isdir(path):
+        return "Cannot read a directory.\n", 400
+    if offset == -1:
+        return jsonify({
+            "data": "",
+            "offset": os.path.getsize(path),
+        })
+    f = open(path)
+    f.seek(offset)
+    length = max_read_length if length == -1 else length
+    if length > max_read_length:
+        return f"Requested length for file read, {length} is greater than max allowed length, {max_read_length}", 400
+    data = f.read(length)
+    f.close()
+    return jsonify({
+        "data": data,
+        "offset": offset,
+    })
+
+
+permission_strings = {
+    '0': '---',
+    '1': '--x',
+    '2': '-w-',
+    '3': '-wx',
+    '4': 'r--',
+    '5': 'r-x',
+    '6': 'rw-',
+    '7': 'rwx',
+}
+# maps permission bit masks to a linux permission string. e.g. 775 -> "rwxrwxr-x"
+permissions = {n - 512: ''.join(permission_strings[c] for c in str(oct(n))[-3:]) for n in range(512, 1024)}
```

Comment:
Can I suggest this function instead:
def perm_bits_to_string(n):
    ''.join([yield ("rwxrwxrwx"[ii] if (n&(1<<(8-ii))!=0) else "-" for ii in range(0,8)])

Reply:
will update